confgenerator/feature.go

```diff
@@ -0,0 +1,210 @@
+package confgenerator
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+	"strings"
+)
+
+// Errors returned by ExtractFeatures can be tested against these errors using
+// errors.Is
+var (
+	ErrTrackingInlineStruct = errors.New("cannot have tracking on inline struct")
+	ErrInvalidType          = errors.New("object in path must be of type Component")
+)
+
+type Feature struct {
+	// Module defines the sub-agent: metrics or logging
+	Module string
+	// Kind defines the kind: receivers or processors
+	Kind string
+	// Type from Component.Type()
+	Type string
+	// Key: set of keys that will be joined together for feature tracking metrics
+	Key []string
+	// Value defined from fields of UnifiedConfig.
+	Value string
+}
+
+// ExtractFeatures fields that containing a tracking tag will be tracked.
+// Automatic collection of bool or int fields. Any value that exists on tracking
+// tag will be used instead of value from UnifiedConfig.
+func ExtractFeatures(uc *UnifiedConfig) ([]Feature, error) {
+	allFeatures := getOverriddenDefaultPipelines(uc)
+	f, err := trackingFeatures(reflect.ValueOf(uc), metadata{}, Feature{})
+	if err != nil {
+		return nil, err
+	}
+	allFeatures = append(allFeatures, f...)
+	return allFeatures, nil
+}
+
+func trackingFeatures(c reflect.Value, m metadata, feature Feature) ([]Feature, error) {
+	t := c.Type()
+
+	if t.Kind() == reflect.Ptr {
+		t = t.Elem()
+	}
+
+	v := reflect.Indirect(c)
+	if v.Kind() == reflect.Invalid || v.IsZero() {
+		return nil, nil
+	}
+
+	var features []Feature
+
+	switch kind := t.Kind(); {
+	case kind == reflect.Struct:
+		if m.IsInline && m.HasTracking {
+			return nil, ErrTrackingInlineStruct
+		}
+		if feature.Module == "" {
+			feature.Module = m.PathName
+		} else if feature.Kind == "" {
+			feature.Kind = m.PathName
+		} else if !m.IsInline {
+			feature.Key = append(feature.Key, m.PathName)
+		}
+
+		// Iterate over all available fields and read the tag value
+		for i := 0; i < t.NumField(); i++ {
+			// Get the field, returns https://golang.org/pkg/reflect/#StructField
+			field := t.Field(i)
```

Comment:
Could you add the contract of the `ExtractFeatures` interface somewhere as a comment for future readers? We should document that the above example is going to be behaving differently. The
```
a nested `yaml:"a", tracking:""`
```
Line above won't be allowed since tracking a field means giving it a key so it can't be empty. The above example produces 5 metrics:

- a
- a.str
- a.in
- b.str
- b.in

Reply:
Also this makes for a good unit test

Reply:
Changed in, unit test added: [237e331](https://github.com/GoogleCloudPlatform/ops-agent/pull/853/commits/237e331081f573c60c5da6d731121d7f4cd56a46)